src/language-js/print/ternary.js

```diff
@@ -153,6 +153,120 @@ function printTernaryTest(path, options, print) {
   return printed;
 }
 
+function isChainElement(node) {
+  return (
+    node.type === "MemberExpression" ||
+    node.type === "OptionalMemberExpression" ||
+    node.type === "CallExpression" ||
+    node.type === "OptionalCallExpression" ||
+    node.type === "TSNonNullExpression"
+  );
+}
+
+const ancestorNameMap = new Map([
+  ["AssignmentExpression", "right"],
+  ["VariableDeclarator", "init"],
+  ["ReturnStatement", "argument"],
+  ["ThrowStatement", "argument"],
+  ["UnaryExpression", "argument"],
+  ["YieldExpression", "argument"],
+]);
+function shouldExtraIndentForConditionalExpression(path) {
+  const node = path.getValue();
+  if (node.type !== "ConditionalExpression") {
+    return false;
+  }
+
+  const parentName = path.callParent((path) => path.getName());
+  if (!parentName) {
+    return false;
+  }
+
+  let ancestorCount = 1;
+  let ancestor = path.getParentNode(ancestorCount);
+  while (isChainElement(ancestor)) {
+    ancestor = path.getParentNode(ancestorCount++);
+  }
+  if (ancestor.type === "BinaryExpression") {
+    return false;
+  }
+  const checkAncestor = (node) => {
+    const name = ancestorNameMap.get(ancestor.type);
+    return ancestor[name] === node;
+  };
+
+  const parent = path.getParentNode();
+
+  /**
+   * foo = (
+   *   condition
+   *     ? first
+   *     : second
+   * ) as SomeType;
+   */
+  if (parent.type === "TSAsExpression" && checkAncestor(parent)) {
+    return true;
+  }
+
+  /**
+   * foo = (
+   *   condition
+   *     ? first
+   *     : second
+   * )(arguments);
+   */
+  if (
+    (parent.type === "CallExpression" ||
+      parent.type === "OptionalCallExpression") &&
+    checkAncestor(parent) &&
+    path.getName() === "callee"
+  ) {
+    return true;
+  }
+
+  /**
+   * foo = new (
+   *   condition
+   *     ? first
+   *     : second
+   * )(arguments);
+   */
+  if (
+    parent.type === "NewExpression" &&
+    checkAncestor(parent) &&
+    path.getName() === "callee"
+  ) {
+    return true;
+  }
+
+  /**
+   * foo = (
+   *   condition
+   *     ? first
+   *     : second
+   * ).member.chaining;
+   */
+  if (
+    parent.type === "MemberExpression" ||
+    parent.type === "OptionalMemberExpression"
+  ) {
+    let memberChainRoot;
+    let count = 0;
+    for (;;) {
```

Comment:
This part doesn't look good, but I can't think a better way, maybe extract to a function will help?

Reply:
I don't like this part too..., I tried to extract utils for chains (52461f2)

Reply:
It can return earlier in loop, and  should take another argument, because it should not check node self and parentNode.